### PR TITLE
CH: Add new keys format

### PR DESCRIPTION
**Pull request by kamel-rahim**
This new format use zset to make listing faster

### CODE REVIEW

File: oioswift/common/middleware/container_hierarchy.py [hunk 1]
```diff
@@ -43,6 +44,106 @@
...
+                 end
+            end
+
+            if #keys <= 1 then
+                finish = true
+            else
+                marker = "[" .. keys[#keys]
+            end
+
+            count = get_size(set)
```
Comment (murlock) [inline oioswift/common/middleware/container_hierarchy.py]: the `get_size` call is maybe redundant with l.126 ?

File: oioswift/common/middleware/container_hierarchy.py [hunk 2]
```diff
@@ -43,6 +44,106 @@
...
+                     local index = string.find(elem, prefix)
+                     if index == nil or index > 1 then
+                         check = false
+                     end
+                 end
+
+                 if check and set[elem] == nil then
+                     count = count + 1
+                     set[elem] = true
+                 end
```
Comment (murlock) [inline oioswift/common/middleware/container_hierarchy.py]: We can imagine an optimization here in a dedicated PR:
If we have lot of keys like below and we asked a non-recursive listing on /
- dir1/dir2/...
- dir1/dir3/...
...
- dir1/dir100000/...
- dir2/...
we could stop the loop and restart maker with "[" .. "dir1/" + \xef\xbf\xbd"
(like https://github.com/open-io/oio-sds/pull/1689/files#diff-3b3bc99683f0c5fe810b2291cc9beeafR1518)

File: oioswift/common/middleware/container_hierarchy.py [hunk 3]
```diff
@@ -43,6 +44,106 @@
...
+            marker = "[" .. marker
+        end
+
+        local nb_elem = redis.call('ZCOUNT', bucket, '-inf', '+inf')
+        local count = 0
+        while count < limit + 2  and finish == false do
+            local list_until = math.min(rank + limit + 2, nb_elem)
+            keys = redis.call('ZRANGEBYLEX', bucket, marker, '+', 'LIMIT',
+                               0, limit)
+            redis.log(redis.LOG_WARNING, " keys = ", keys)
```
[the remark this answers is not in the document]
Comment (kamel-rahim) [reply oioswift/common/middleware/container_hierarchy.py]: Sorry this was debug log they should be removed

File: oioswift/common/middleware/container_hierarchy.py [hunk 4]
```diff
@@ -43,6 +44,106 @@
...
+        local rank = 0
+        local finish = false
+
+        if marker == "" then
+            marker = "-"
+        else
+            marker = "[" .. marker
+        end
+
+        local nb_elem = redis.call('ZCOUNT', bucket, '-inf', '+inf')
```
[the remark this answers is not in the document]
Comment (kamel-rahim) [reply oioswift/common/middleware/container_hierarchy.py]: Yes this is code I should have removed.
It come from the old listing.
I used before ZRANGE and I changed it to ZRANGEBYLEX but forgot to remove the old code

File: oioswift/common/middleware/container_hierarchy.py (hunk 1 above)
Comment (kamel-rahim) [reply oioswift/common/middleware/container_hierarchy.py]: get_size is for the set that will be returned (no more than `limit` unique elements).
keys are the elements that came from ZRANGEBYLEX.

File: oioswift/common/middleware/container_hierarchy.py [hunk 5]
```diff
@@ -192,8 +193,11 @@ def hset(self, key, path, val):
     def hsetnx(self, key, path, val):
         return self.conn.hsetnx(key, path, val)
 
-    def zsetnx(self, key, path):
-        return self.conn.zadd(key, {path: 1}, nx=True)
+    def zset(self, key, path):
```
Comment (murlock) [inline oioswift/common/middleware/container_hierarchy.py]: Maybe we should compute once LooseVersion in `__init__` method
or having `zset_legacy` `zset3` and in `__init__`:
```
if  if LooseVersion(self.__redis_mod.__version__) < LooseVersion("3.0.0"):
    self.zset = self.zset_legacy
else:
    self.zset = self.zset3
```

**Pull request merged**